beetsplug/web/subsonic.py

```diff
@@ -0,0 +1,666 @@
+# -*- coding: utf-8 -*-
+# This file is part of beets.
+# Copyright 2016, Joël Charles.
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+
+"""
+A (subset of) subsonic API implementation on top of Web plugin.
+See http://www.subsonic.org/pages/api.jsp
+"""
+
+from __future__ import division, absolute_import, print_function
+
+import datetime
+import mimetypes
+import os
+import re
+from time import time
+from xml.dom.minidom import Text
+
+import six
+from flask import send_file, g, Response, Blueprint, request
+from flask.json import jsonify
+
+from beets.util.random import random_objs
+
+DEFAULT_VERSION = '1.10.1'
+subsonic_routes = Blueprint('subsonic', 'subsonic')
+
+
+def dict2xml(d, root_node=None):
+    """Transform python dict to XML string representation"""
+    wrap = False if None is root_node or isinstance(d, list) else True
+    root = 'objects' if None is root_node else root_node
+    root_singular = root[:-1] if 's' == root[-1] and \
+                                 None is root_node else root
+    xml = ''
+    children = []
+
+    if isinstance(d, dict):
+        for key, value in dict.items(d):
+            if isinstance(value, dict):
+                children.append(dict2xml(value, key))
+            elif isinstance(value, list):
+                children.append(dict2xml(value, key))
+            else:
+                t = Text()
+                if isinstance(value, bool):
+                    t.data = str(value).lower()
+                elif isinstance(value, bytes):
+                    t.data = value.decode('utf-8')
+                elif not isinstance(value, six.text_type):
+                    t.data = six.text_type(value)
+                else:
+                    t.data = value
+                xml = u'{} {}="{}"'.format(xml, key, t.toxml())
+    else:
+        for value in d:
+            children.append(dict2xml(value, root_singular))
+
+    end_tag = u'>' if 0 < len(children) else u'/>'
+
+    if wrap or isinstance(d, dict):
+        xml = u"<{}{}{}".format(root, xml, end_tag)
+
+    if 0 < len(children):
+        for child in children:
+            xml = u"{}{}".format(xml, child)
+
+        if wrap or isinstance(d, dict):
+            xml = u"{}</{}>".format(xml, root)
+
+    return xml
```

Comment:
Why all the string building instead of using the built-in XML builder in the standard library?

(I'm not a project maintainer, just jumping on here because I am also working with and potentially modifying the web API.)